sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,662 @@
+
+	/**
+	 * Set value of a GPIO output.
+	 */
+	inline void set_gpio_output_bit(GpioPin pin, bool value) const
+	{
+		uint32_t shift  = static_cast<uint8_t>(pin);
+		uint32_t output = gpio()->output;
+		output &= ~(1 << shift);
+		output |= value << shift;
+		gpio()->output = output;
+	}
+
+	/**
+	 * Read a register from the KSZ8851.
+	 */
+	uint16_t register_read(RegisterOffset reg) const
+	{
+		uint8_t addr = static_cast<uint8_t>(reg);
+		// KSZ8851 have 16-bit registers but accesses are 32 bits,
+		// so use byte enable to access within each 32-bit word.
+		uint8_t be = (addr & 0x2) == 0 ? 0b0011 : 0b1100;
+		uint8_t bytes[2];
+		bytes[0] = (0b00 << 6) | (be << 2) | (addr >> 6);
+		bytes[1] = (addr << 2) & 0b11110000;
+
+		set_gpio_output_bit(GpioPin::EthCsPin, false);
+		spi()->blocking_write(bytes, 2);
+		uint16_t val;
+		spi()->blocking_read((uint8_t *)&val, 2);
+		set_gpio_output_bit(GpioPin::EthCsPin, true);
+		return val;
+	}
+
+	/**
+	 * Write a register to KSZ8851.
+	 */
+	void register_write(RegisterOffset reg, uint16_t val) const
+	{
+		uint8_t addr = static_cast<uint8_t>(reg);
+		uint8_t be   = (addr & 0x2) == 0 ? 0b0011 : 0b1100;
+		uint8_t bytes[2];
+		bytes[0] = (0b01 << 6) | (be << 2) | (addr >> 6);
+		bytes[1] = (addr << 2) & 0b11110000;
+
+		set_gpio_output_bit(GpioPin::EthCsPin, false);
+		spi()->blocking_write(bytes, 2);
+		spi()->blocking_write((uint8_t *)&val, 2);
+		spi()->wait_idle();
+		set_gpio_output_bit(GpioPin::EthCsPin, true);
+	}
+
+	void register_set(RegisterOffset reg, uint16_t mask) const
+	{
+		uint16_t old = register_read(reg);
+		register_write(reg, old | mask);
+	}
+
+	void register_clear(RegisterOffset reg, uint16_t mask) const
+	{
+		uint16_t old = register_read(reg);
+		register_write(reg, old & ~mask);
+	}
+
+	/**
+	 * Helper.  Returns a pointer to the SPI device.
+	 */
+	[[nodiscard, gnu::always_inline]] Capability<volatile SonataSpi> spi() const
+	{
+		return MMIO_CAPABILITY(SonataSpi, spi2);
+	}
+
+	/**
+	 * Helper.  Returns a pointer to the GPIO device.
+	 */
+	[[nodiscard, gnu::always_inline]] Capability<volatile SonataGPIO>
+	gpio() const
+	{
+		return MMIO_CAPABILITY(SonataGPIO, gpio);
+	}
+
+	/**
+	 * Number of frames yet to be received since last interrupt acknowledgement.
+	 */
+	uint16_t frames_to_process = 0;
+
+	public:
+	/**
+	 * Initialise a reference to the Ethernet device.
+	 */
+	Ksz8851Ethernet()
+	{
+		// Reset chip. It needs to be hold in reset for at least 10ms.
+		set_gpio_output_bit(GpioPin::EthRstPin, false);
+		thread_millisecond_wait(20);
+		set_gpio_output_bit(GpioPin::EthRstPin, true);
+
+		uint16_t cider = register_read(RegisterOffset::CIDER);
+		Debug::log("Chip ID is {}", cider);
+
+		// Check the chip ID. The last nibble is revision ID and can be ignored.
+		Debug::Assert((cider & 0xFFF0) == 0x8870, "Unexpected Chip ID");
+
+		// Enable QMU Transmit Frame Data Pointer Auto Increment
+		register_write(RegisterOffset::TXFDPR, 0x4000);
```

Comment:
There are a lot of magic numbers here (0x4000, 0x00EE, and so on) that would benefit from being symbolic.